cwe_checker_rs/src/checkers/cwe_78/context/mod.rs

```diff
@@ -0,0 +1,563 @@
+                        ))
+                        .unwrap(),
+                )),
+            };
+            if let Ok(address) = pi_state.eval(&address_exp) {
+                new_state.save_taint_to_memory(&address, Taint::Tainted(size));
+            }
+        }
+        new_state
+    }
+
+    // Iterates over the given parameters of a function and returns an updated state
+    pub fn taint_parameters(
+        &self,
+        state: &State,
+        parameters: Vec<Arg>,
+        call_source_node: NodeIndex,
+    ) -> State {
+        let mut new_state = state.clone();
+        for parameter in parameters {
+            match parameter {
+                Arg::Register(param) => {
+                    new_state.set_register_taint(&param, Taint::Tainted(param.size))
+                }
+                Arg::Stack { offset, size } => {
+                    new_state = self.taint_stack_parameters(
+                        new_state.clone(),
+                        call_source_node,
+                        offset,
+                        size,
+                    );
+                }
+            }
+        }
+
+        new_state
+    }
+
+    /// Checks whether the current def term is the last def term
+    /// of its corresponding block and if so, returns the block term and correspondoing
+    pub fn get_blk_start_node_if_last_def(&self, def: &Term<Def>) -> Option<NodeIndex> {
+        if let Some(sub) = self.current_sub {
+            if let Some(node) = self
+                .block_start_node_map
+                .get(&(def.tid.clone(), sub.tid.clone()))
+            {
+                return Some(*node);
+            }
+        } else {
+            panic!("Missing current Sub.");
+        }
+
+        None
+    }
+
+    /// Creates a map from def terms to their corresponding pointer inference states
+    /// by taking the pointer inference state of the BlkStart node and updating it
+    /// for each def term in the block
+    pub fn create_pi_def_map(
+        &self,
+        block_start_node: NodeIndex,
+    ) -> Option<HashMap<Tid, PointerInferenceState>> {
+        if let Some(block_node) = self.get_pi_graph().node_weight(block_start_node) {
+            if let Some(pi_value) = self
+                .pointer_inference_results
+                .get_node_value(block_start_node)
+            {
+                let mut pi_def_map: HashMap<Tid, PointerInferenceState> = HashMap::new();
+
+                let pi_context = self.pointer_inference_results.get_context();
+                let mut new_pi_state = Some(pi_value.unwrap_value().clone());
+
+                for def in block_node.get_block().term.defs.iter() {
+                    // Add the pi state to the map that is available after the def was executed
+                    // If no state is available after the update_def() call, none is added
+                    new_pi_state = pi_context.update_def(&new_pi_state.unwrap(), def);
```

Comment:
If update_def returns `None` before the last def of the block, then the `new_pi_state.unwrap()` would panic during the next iteration. We should `break` out of the loop before that happens.